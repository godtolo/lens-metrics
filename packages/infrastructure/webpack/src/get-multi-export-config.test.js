import getMultiExportConfig from "./get-multi-export-config";

describe("get-multi-export-config", () => {
  let actual;
  let maximalPackageJson;

  beforeEach(() => {
    maximalPackageJson = {
      name: "some-name",

      lensMultiExportConfig: {
        ".": {
          buildType: "node",
          entrypoint: "./index.ts",
        },

        "./some-entrypoint": {
          buildType: "node",
          entrypoint: "./some-entrypoint/index.ts",
        },

        "./some-other-entrypoint": {
          buildType: "react",
          entrypoint: "./some-other-entrypoint/index.ts",
        },
      },

      exports: {
        ".": {
          types: "./dist/index.d.ts",
          require: "./dist/index.js",
          import: "./dist/index.js",
          default: "./dist/index.js",
        },
        "./some-entrypoint": {
          types: "./dist/some-entrypoint/index.d.ts",
          require: "./dist/some-entrypoint/index.js",
          import: "./dist/some-entrypoint/index.js",
          default: "./dist/some-entrypoint/index.js",
        },
        "./some-other-entrypoint": {
          types: "./dist/some-other-entrypoint/index.d.ts",
          require: "./dist/some-other-entrypoint/index.js",
          import: "./dist/some-other-entrypoint/index.js",
          default: "./dist/some-other-entrypoint/index.js",
        },
      },
    };
  });

  it("given maximal package.json, when creating configuration, works", () => {
    actual = getMultiExportConfig(maximalPackageJson, {
      nodeConfig: nodeConfigStub,
      reactConfig: reactConfigStub,
    });

    expect(actual).toEqual([
      {
        name: "./index.ts",
        stub: "node",
        entry: { index: "./index.ts" },
        output: { some: "value", path: "/some-build-directory" },
      },

      {
        name: "./some-entrypoint/index.ts",
        stub: "node",
        entry: { index: "./some-entrypoint/index.ts" },

        output: {
          some: "value",
          path: "/some-build-directory/some-entrypoint",
        },
      },

      {
        name: "./some-other-entrypoint/index.ts",
        stub: "react",
        entry: { index: "./some-other-entrypoint/index.ts" },

        output: {
          some: "other-value",
          path: "/some-build-directory/some-other-entrypoint",
        },
      },
    ]);
  });

  it("given maximal package.json but path for entrypoint in exports do not match output, when creating configuration, throws", () => {
    maximalPackageJson.exports["./some-entrypoint"].default = "wrong-path";

    expect(() => {
      getMultiExportConfig(maximalPackageJson, {
        nodeConfig: nodeConfigStub,
        reactConfig: reactConfigStub,
      });
    }).toThrow(
      'Tried to get multi export config but exports of package.json for "some-name" did not match exactly:'
    );
  });

  it("given maximal package.json but exports do not match lens multi export config, when creating configuration, throws", () => {
    maximalPackageJson.exports = {};

    expect(() => {
      getMultiExportConfig(maximalPackageJson, {
        nodeConfig: nodeConfigStub,
        reactConfig: reactConfigStub,
      });
    }).toThrow(
      'Tried to get multi export config but exports of package.json for "some-name" did not match exactly:'
    );
  });

  it("given maximal package.json but exports are missing, when creating configuration, throws", () => {
    delete maximalPackageJson.exports;

    expect(() => {
      getMultiExportConfig(maximalPackageJson, {
        nodeConfig: nodeConfigStub,
        reactConfig: reactConfigStub,
      });
    }).toThrow(
      'Tried to get multi export config but exports of package.json for "some-name" did not match exactly:'
    );
  });

  it("given maximal package.json but lens multi export config is missing, when creating configuration, throws", () => {
    delete maximalPackageJson.lensMultiExportConfig;

    expect(() => {
      getMultiExportConfig(maximalPackageJson, {
        nodeConfig: nodeConfigStub,
        reactConfig: reactConfigStub,
      });
    }).toThrow(
      'Tried to get multi export config for package "some-name" but configuration is missing.'
    );
  });

  it("given maximal package.json but a build type is incorrect, when creating configuration, throws", () => {
    maximalPackageJson.lensMultiExportConfig["."].buildType = "some-invalid";

    expect(() => {
      getMultiExportConfig(maximalPackageJson, {
        nodeConfig: nodeConfigStub,
        reactConfig: reactConfigStub,
      });
    }).toThrow(
      'Tried to get multi export config for package "some-name" but build types "some-invalid" were not any of "node", "react".'
    );
  });

  it("given maximal package.json but entrypoint is missing, when creating configuration, throws", () => {
    delete maximalPackageJson.lensMultiExportConfig["./some-entrypoint"]
      .entrypoint;

    expect(() => {
      getMultiExportConfig(maximalPackageJson, {
        nodeConfig: nodeConfigStub,
        reactConfig: reactConfigStub,
      });
    }).toThrow(
      'Tried to get multi export config for package "some-name" but entrypoint was missing for "./some-entrypoint".'
    );
  });
});

const nodeConfigStub = {
  stub: "node",
  output: {
    some: "value",
    path: "/some-build-directory",
  },
};

const reactConfigStub = {
  stub: "react",

  output: {
    some: "other-value",
    path: "/some-build-directory",
  },
};
