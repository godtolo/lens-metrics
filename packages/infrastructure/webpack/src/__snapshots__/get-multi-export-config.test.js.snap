// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-multi-export-config given maximal package.json, when creating configuration works 1`] = `
Array [
  Object {
    "entry": Object {
      "index": "./index.ts",
    },
    "externals": Array [
      [Function],
      [Function],
    ],
    "externalsPresets": Object {
      "node": true,
    },
    "mode": "production",
    "module": Object {
      "rules": Array [
        Object {
          "loader": "ts-loader",
          "test": /\\\\\\.ts\\(x\\)\\?\\$/,
        },
      ],
    },
    "name": "./index.ts",
    "node": Object {
      "__dirname": true,
      "__filename": true,
    },
    "output": Object {
      "filename": [Function],
      "libraryTarget": "commonjs2",
      "path": "/some-working-directory/dist",
    },
    "performance": Object {
      "hints": "error",
      "maxEntrypointSize": 100000,
    },
    "plugins": Array [
      ForkTsCheckerWebpackPlugin {
        "options": Object {
          "typescript": Object {
            "configOverwrite": Object {
              "compilerOptions": Object {
                "declaration": true,
                "declarationDir": "/some-working-directory/dist",
              },
              "include": Array [
                "./index.ts",
              ],
            },
            "mode": "write-dts",
          },
        },
      },
    ],
    "resolve": Object {
      "extensions": Array [
        ".ts",
        ".tsx",
      ],
    },
    "target": "node",
  },
  Object {
    "entry": Object {
      "index": "./some-entrypoint/index.ts",
    },
    "externals": Array [
      [Function],
      [Function],
    ],
    "externalsPresets": Object {
      "node": true,
    },
    "mode": "production",
    "module": Object {
      "rules": Array [
        Object {
          "loader": "ts-loader",
          "test": /\\\\\\.ts\\(x\\)\\?\\$/,
        },
      ],
    },
    "name": "./some-entrypoint/index.ts",
    "node": Object {
      "__dirname": true,
      "__filename": true,
    },
    "output": Object {
      "filename": [Function],
      "libraryTarget": "commonjs2",
      "path": "/some-working-directory/dist/some-entrypoint",
    },
    "performance": Object {
      "hints": "error",
      "maxEntrypointSize": 100000,
    },
    "plugins": Array [
      ForkTsCheckerWebpackPlugin {
        "options": Object {
          "typescript": Object {
            "configOverwrite": Object {
              "compilerOptions": Object {
                "declaration": true,
                "declarationDir": "/some-working-directory/dist/some-entrypoint",
              },
              "include": Array [
                "./some-entrypoint/index.ts",
              ],
            },
            "mode": "write-dts",
          },
        },
      },
    ],
    "resolve": Object {
      "extensions": Array [
        ".ts",
        ".tsx",
      ],
    },
    "target": "node",
  },
  Object {
    "entry": Object {
      "index": "./some-other-entrypoint/index.ts",
    },
    "externals": Array [
      [Function],
      [Function],
    ],
    "externalsPresets": Object {
      "node": true,
    },
    "mode": "production",
    "module": Object {
      "rules": Array [
        Object {
          "loader": "ts-loader",
          "test": /\\\\\\.ts\\(x\\)\\?\\$/,
        },
        Object {
          "test": /\\\\\\.s\\?css\\$/,
          "use": Array [
            Object {
              "some": "miniCssExtractPluginLoader",
            },
            Object {
              "loader": "css-loader",
              "options": Object {
                "modules": Object {
                  "auto": /\\\\\\.module\\\\\\./i,
                  "localIdentName": "[name]__[local]--[hash:base64:5]",
                  "mode": "local",
                },
                "sourceMap": false,
              },
            },
            Object {
              "loader": "sass-loader",
              "options": Object {
                "sourceMap": false,
              },
            },
          ],
        },
      ],
    },
    "name": "./some-other-entrypoint/index.ts",
    "node": Object {
      "__dirname": true,
      "__filename": true,
    },
    "output": Object {
      "filename": [Function],
      "libraryTarget": "commonjs2",
      "path": "/some-working-directory/dist/some-other-entrypoint",
    },
    "performance": Object {
      "hints": "error",
      "maxEntrypointSize": 100000,
    },
    "plugins": Array [
      ForkTsCheckerWebpackPlugin {
        "options": Object {
          "typescript": Object {
            "configOverwrite": Object {
              "compilerOptions": Object {
                "declaration": true,
                "declarationDir": "/some-working-directory/dist/some-other-entrypoint",
              },
              "include": Array [
                "./some-other-entrypoint/index.ts",
              ],
            },
            "mode": "write-dts",
          },
        },
      },
      MiniCssExtractPlugin {
        "_sortedModulesCache": WeakMap {},
        "options": Object {
          "chunkFilename": "[name].css",
          "experimentalUseImportModule": undefined,
          "filename": "[name].css",
          "ignoreOrder": false,
          "runtime": true,
        },
        "runtimeOptions": Object {
          "attributes": undefined,
          "insert": undefined,
          "linkType": "text/css",
        },
      },
    ],
    "resolve": Object {
      "extensions": Array [
        ".ts",
        ".tsx",
      ],
    },
    "target": "node",
  },
]
`;
